### PR TITLE
Obsidian connect

### PR DESCRIPTION
**Pull request by edewit**
- [ ] Have you created an [issue](https://github.com/redhat-kontinuity/catapult/issues) and used it in the commit message?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/redhat-kontinuity/catapult/pulls) for the same issue?
- [x] Have you built the project locally prior to submission with `mvn clean install`?

-----

Fixed issue #158 adds an endpoint under `api/catapult/upload` that accepts an zip project file and creates an github repo for it and a openshift project

### CODE REVIEW

File: core/core-impl/src/main/java/org/kontinuity/catapult/core/impl/CatapultImpl.java [hunk 1]
```diff
@@ -44,13 +57,30 @@ public Boom fling(final Projectile projectile) throws IllegalArgumentException {
...
+        final GitHubRepository gitHubRepository;
+        if (projectile instanceof ForkProjectile) {
+            // Get properties
+            final String sourceRepoName = ((ForkProjectile) projectile).getSourceGitHubRepo();
+            gitHubRepository = gitHubService.fork(sourceRepoName);
+        } else {
+            gitHubRepository = gitHubService.createRepository(projectile.getOpenShiftProjectName(), "quickstart");
+            String projectLocation = ((CreateProjectile) projectile).getProjectLocation();
+            try {
+                Git repo = Git.init().setDirectory(new File(projectLocation)).call();
```
Comment (gastaldi) [inline core/core-impl/src/main/java/org/kontinuity/catapult/core/impl/CatapultImpl.java]: Don't forget to close this repo object (use a try-with-resources block)

File: web/src/main/java/org/kontinuity/catapult/web/api/CatapultResource.java [hunk 2]
```diff
@@ -122,4 +186,108 @@ public Response fling(
...
+      // came back from GitHub oath already unpacked the zip file just fling it.
+      ProjectileOrResponse result = createProjectileForCreateRequest(request, projectLocation);
+      if (result.hasResponse()) {
+         return result.getResponse();
+      }
+
+      return fling(result.getProjectile());
+   }
+
+   private void unzip(File zipFile) throws IOException {
```
Comment (gastaldi) [inline web/src/main/java/org/kontinuity/catapult/web/api/CatapultResource.java]: I think this method should be somewhere else, not this class

File: web/src/main/java/org/kontinuity/catapult/web/api/CatapultResource.java [hunk 3]
```diff
@@ -122,4 +186,108 @@ public Response fling(
...
+         if ((filename.trim().startsWith("filename"))) {
+            String[] name = filename.split("=");
+            return sanitizeFilename(name[1]);
+         }
+      }
+      return "unknown";
+   }
+
+   private String sanitizeFilename(String s) {
+      return s.trim().replaceAll("\"", "");
```
Comment (gastaldi) [inline web/src/main/java/org/kontinuity/catapult/web/api/CatapultResource.java]: I think this method should be somewhere else, not this class

File: core/core-api/src/main/java/org/kontinuity/catapult/core/api/CreateProjectileBuilder.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+ * <li>projectLocation</li>
+ * </ul>
+ * <p>
+ * Each property's valid value and purpose is documented in its setter method.
+ */
+public class CreateProjectileBuilder extends ProjectileBuilder {
+   private String projectLocation;
+
+   CreateProjectileBuilder(String gitHubAccessToken, String openShiftProjectName) {
+      super(gitHubAccessToken, openShiftProjectName);
```
Comment (gastaldi) [inline core/core-api/src/main/java/org/kontinuity/catapult/core/api/CreateProjectileBuilder.java]: is this openshiftProjectName or projectLocation?

File: core/core-api/src/main/java/org/kontinuity/catapult/core/api/ProjectileBuilder.java [hunk 5]
```diff
@@ -1,189 +1,125 @@
...
    /**
-    * @return the path to the file that contains the template to apply on the OpenShift project.
+    * @return The name to use in creating the new OpenShift project
     */
-   public String getPipelineTemplatePath() { return this.pipelineTemplatePath; }
+   public String getOpenShiftProjectName() {
+      return openShiftProjectName;
+   }
+
+   public CreateProjectileBuilder createTye() {
```
Comment (gastaldi) [inline core/core-api/src/main/java/org/kontinuity/catapult/core/api/ProjectileBuilder.java]: Typo here

File: core/core-impl/src/main/java/org/kontinuity/catapult/core/impl/CatapultImpl.java [hunk 6]
```diff
@@ -61,37 +90,43 @@ public Boom fling(final Projectile projectile) throws IllegalArgumentException {
...
+                final StringBuilder sb = new StringBuilder();
+                sb.append("https://raw.githubusercontent.com/");
+                sb.append(forkProjectile.getSourceGitHubRepo());
+                sb.append('/');
+                sb.append(forkProjectile.getGitRef());
+                sb.append('/');
+                sb.append(forkProjectile.getPipelineTemplatePath());
+                final URI pipelineTemplateUri;
+                try {
+                    pipelineTemplateUri = new URI(sb.toString());
```
Comment (gastaldi) [inline core/core-impl/src/main/java/org/kontinuity/catapult/core/impl/CatapultImpl.java]: Can we use UriBuilder in here?

**Comment by ALRubinger** [comment]
From meeting: prereq to PR accepted is integration test using the new endpoint

**Comment by edewit** [comment]
@ALRubinger added IT test

**Comment by ALRubinger** [comment]
Rebase on master, now seeing when running `mvn clean install -Pit` (make sure you have MiniShift installed and running, and env vars set as noted in README:

```
-------------------------------------------------------------------------------
Test set: org.kontinuity.catapult.core.impl.CatapultIT
-------------------------------------------------------------------------------
Tests run: 2, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 89.781 sec <<< FAILURE! - in org.kontinuity.catapult.core.impl.CatapultIT
flingCreate(org.kontinuity.catapult.core.impl.CatapultIT)  Time elapsed: 7.812 sec  <<< ERROR!
java.lang.RuntimeException: Could not remove RedHatDistortion/jboss-eap-quickstarts
	at org.kontinuity.catapult.core.impl.CatapultIT.reset(CatapultIT.java:99)
Caused by: java.io.FileNotFoundException: Failed to delete RedHatDistortion/jboss-eap-quickstarts; might not exist, or you might need the delete_repo scope in your token: http://stackoverflow.com/a/19327004/12916
	at org.kontinuity.catapult.core.impl.CatapultIT.reset(CatapultIT.java:99)
Caused by: java.io.FileNotFoundException: {"message":"Not Found","documentation_url":"https://developer.github.com/v3"}
	at org.kontinuity.catapult.core.impl.CatapultIT.reset(CatapultIT.java:99)
Caused by: java.io.FileNotFoundException: https://api.github.com/repos/RedHatDistortion/jboss-eap-quickstarts
	at org.kontinuity.catapult.core.impl.CatapultIT.reset(CatapultIT.java:99)
```

**Comment by ALRubinger** [comment]
My rebase in progress: https://github.com/ALRubinger/catapult/commit/34cd67884443b9c0832320d38d06c7d9a6a7fa32

**Comment by edewit** [comment]
@ALRubinger is that because there is no github repo called https://github.com/RedHatDistortion/jboss-eap-quickstarts?

File: web/src/main/java/org/kontinuity/catapult/web/api/CatapultResource.java [hunk 7]
```diff
@@ -118,8 +168,80 @@ public Response fling(
             log.finest("Redirect issued to: " + consoleOverviewUri.toString());
          }
       } catch (final URISyntaxException urise) {
-         return Response.serverError().entity(urise).build();
+         throw new WebApplicationException("couldn't get console location do you have the environment variable set", urise);
```
Comment (gastaldi) [inline web/src/main/java/org/kontinuity/catapult/web/api/CatapultResource.java]: What environment variable?

File: core/core-api/src/main/java/org/kontinuity/catapult/core/api/ProjectileBuilder.java [hunk 8]
```diff
@@ -1,189 +1,125 @@
...
-        return this.gitHubAccessToken;
-    }
-
    /**
-    * @return The name to use in creating the new OpenShift project
+    * @return the GitHub access token we have obtained from the user as part of
+    * the OAuth process
     */
-   public String getOpenShiftProjectName() { return openShiftProjectName;  }
+   public String getGitHubAccessToken() {
```
Comment (gastaldi) [inline core/core-api/src/main/java/org/kontinuity/catapult/core/api/ProjectileBuilder.java]: I think we should get rid of these getter methods in the Builder. They are already available in the built Projectile

File: core/core-impl/src/main/java/org/kontinuity/catapult/core/impl/CatapultImpl.java [hunk 9]
```diff
@@ -44,13 +58,29 @@ public Boom fling(final Projectile projectile) throws IllegalArgumentException {
...
+        } else {
+            gitHubRepository = gitHubService.createRepository(projectile.getOpenShiftProjectName(), "quickstart");
+            String projectLocation = ((CreateProjectile) projectile).getProjectLocation();
+            try (Git repo = Git.init().setDirectory(new File(projectLocation)).call()) {
+                repo.add().addFilepattern(".").call();
+                repo.commit().setMessage("initial version").call();
+                RemoteAddCommand add = repo.remoteAdd();
+                add.setName("origin");
+                add.setUri(new URIish(gitHubRepository.getGitCloneUri().toURL()));
+                add.call();
```
Comment (gastaldi) [inline core/core-impl/src/main/java/org/kontinuity/catapult/core/impl/CatapultImpl.java]: These methods can be chained afaik

**Pull request closed without merging**